Utilities/build-script-helper.py

```diff
@@ -8,230 +8,306 @@
-
-def handle_invocation(swift_exec, args):
-  swiftpm_args = get_swiftpm_options(swift_exec, args)
-
-  env = os.environ
-  # Set the toolchain used in tests at runtime
-  env['SOURCEKIT_TOOLCHAIN_PATH'] = args.toolchain
-  env['INDEXSTOREDB_TOOLCHAIN_BIN_PATH'] = args.toolchain
-  # Use local dependencies (i.e. checked out next sourcekit-lsp).
-  if not args.no_local_deps:
-    env['SWIFTCI_USE_LOCAL_DEPS'] = "1"
-
-  if args.ninja_bin:
-    env['NINJA_BIN'] = args.ninja_bin
-
-  if args.sanitize and 'address' in args.sanitize:
-    # Workaround reports in Foundation: https://bugs.swift.org/browse/SR-12551
-    env['ASAN_OPTIONS'] = 'detect_leaks=false'
-  if args.sanitize and 'undefined' in args.sanitize:
-    supp = os.path.join(args.package_path, 'Utilities', 'ubsan_supressions.supp')
-    env['UBSAN_OPTIONS'] = 'halt_on_error=true,suppressions=%s' % supp
-  if args.sanitize and 'thread' in args.sanitize:
-    env['TSAN_OPTIONS'] = 'halt_on_error=true'
-
-  if not args.no_clean:
-    print('Cleaning ' + args.build_path)
-    shutil.rmtree(args.build_path, ignore_errors=True)
-
-  env['SWIFT_EXEC'] = '%sc' % (swift_exec)
-
-  if args.action == 'build':
-    swiftpm('build', swift_exec, swiftpm_args, env)
-  elif args.action == 'test':
-    if not args.skip_long_tests:
-      env['SOURCEKIT_LSP_ENABLE_LONG_TESTS'] = '1'
+        swiftpm_args += [
+            # Dispatch headers
+            '-Xcxx', '-I', '-Xcxx',
+            os.path.join(args.toolchain, 'lib', 'swift'),
+            # For <Block.h>
+            '-Xcxx', '-I', '-Xcxx',
+            os.path.join(args.toolchain, 'lib', 'swift', 'Block'),
+        ]
+
+    if 'ANDROID_DATA' in os.environ or (args.cross_compile_host and re.match(
+        'android-', args.cross_compile_host)):
+        swiftpm_args += [
+            '-Xlinker', '-rpath', '-Xlinker', '$ORIGIN/../lib/swift/android',
+            # SwiftPM will otherwise try to compile against GNU strerror_r on
+            # Android and fail.
+            '-Xswiftc', '-Xcc', '-Xswiftc', '-U_GNU_SOURCE',
+        ]
+    elif platform.system() == 'Linux':
+        # Library rpath for swift, dispatch, Foundation, etc. when installing
+        swiftpm_args += [
+            '-Xlinker', '-rpath', '-Xlinker', '$ORIGIN/../lib/swift/linux',
+        ]
+
+    build_target = get_build_target(swift_exec, args)
+    if args.cross_compile_host:
+        if build_target == 'x86_64-apple-macosx' and args.cross_compile_host == "macosx-arm64":
+            swiftpm_args += ["--arch", "x86_64", "--arch", "arm64"]
+        elif re.match('android-', args.cross_compile_host):
+            print('Cross-compiling for %s' % args.cross_compile_host)
+            swiftpm_args += ['--destination', args.cross_compile_config]
+        else:
+            fatal_error("cannot cross-compile for %s" % args.cross_compile_host)
+
+    return swiftpm_args
+
+
+def get_swiftpm_environment_variables(swift_exec: str, args: argparse.Namespace) -> Dict[str, str]:
```

Comment:
What do you think of instead having `get_swiftpm_options` return a tuple of arguments and environment? My thinking is that you always need both to be correct.

Reply:
I actually like to keep them separate. That way each method is easier to reason about on its own. Otherwise the unified function becomes larger again and you don’t have a clear separation between the part that sets environments and the part that creates the command line arguments.

Reply:
My concern is that you can forget to use the env variables at all, especially when things like check_call have default values for `env`.  But I don't feel strongly enough about it to block this PR if you disagree.

Reply:
I’ll just remove the default value for `env` from `check_call`.